scipy/cluster/hierarchy.py

```diff
@@ -2256,7 +2257,7 @@ def is_valid_linkage(Z, warning=False, throw=False, name=None):
         ):
             raise ValueError(f'Linkage {name_str}uses non-singleton cluster before'
                              ' it is formed.')
-        if xp.unique_values(Z[:, :2]).shape[0] < n * 2:
+        if xpx.nunique(Z[:, :2]) < n * 2:
```

Comment:
This helps with nan-shaped Dask arrays but doesn't otherwise prevent the issue of having to materialize the graph.
Will iterate more on it.